[lib] handling nulls

dweepgogia opened the pull request:
- For reporting warnings and errors in case of optional and mandatory fields missing, respectively.

dweepgogia committed:
[lib] Handling mandatory/optional fields and reporting custom errors

- For reporting warnings and errors in case of optional and mandatory fields missing, respectively.
- To enable users to extract more information from an error/warning.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
+		fields := strings.Split(tag, ",")
+		omitEmptyTagPresent := contains(fields, "omitempty")
 
+		if tag == "" {
```
Please explain in a comment why this is needed

dweepgogia replied:
Done.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
+// Return true if the string is present in the slicec.
+func contains(a []string, x string) bool {
+	for _, n := range a {
+		if strings.ToLower(x) == strings.ToLower(n) {
```
why use `.ToLower()` here?

dweepgogia replied:
I want it to ignore case while comparing the strings.

dweepgogia replied:
I could also use `strings.EqualFold` for that if you think that's more appropriate.

gallettilance replied:
I understand but is there a situation where `omitempty` has upper case letters?

dweepgogia replied:
Not for `omitempty`, but I was thinking of using this helper in some other part of the code. Although, that did not work out, I decided to make it more general so it can be used in other parts.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
+	case reflect.Struct:
+		return checkStructHasEmptyFields(fieldValue, []string{})
+	default:
+		return []string{"Error: input file is not aligned with Operator-Lifecycle-Manager's CSV type."}
```
suggestion: `Error: input file is not a valid CSV`

dweepgogia replied:
Done

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
+	return csv, nil
+}
+
+func hasEmptyField(val interface{}) []string {
```
The function name implies you are returning a `bool`. Please change the name to better match its function

dweepgogia replied:
Done.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,12 +20,32 @@ func Verify(yamlFileName string) error {
+
+	// Contains error logs for all missing optional and mandatory fields.
+	log := hasEmptyField(csv)
+	sort.SliceStable(log, func(i, j int) bool {
```
Why are you doing this?

dweepgogia replied:
To print the `Warnings` followed by `Errors` separately.

gallettilance replied:
ah I see. Nice touch but I am not sure if this is needed. We should just display them in the order in which they appear for now. We can always change later based on feedback

dweepgogia replied:
Sure.

dweepgogia replied:
Reverted back, based on Eric's feedback.

dweepgogia replied:
Now that we have separate objects for errors and warnings, we do not need to explicitly sort them. So the order of fields is maintained and they are segregated as well :)

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
+// Recursive function that traverses a nested struct passed in as reflect value, and reports for errors/warnings
+// in case of null struct field values.
+// Returns a log of errors as slice of strings.
+func checkStructHasEmptyFields(v reflect.Value, log []string) []string {
```
This function name could be improved. I'd suggest something along the lines of `csvReport` or `csvInspect`

dweepgogia replied:
How about I change `hasEmptyField` to `csvInspect` and `checkStructHasEmptyFields` to `getMissingFields` or `getEmptyFields` because `checkStructHasEmptyFields` has nothing to do with `csv` in general?

gallettilance replied:
Great point! I agree with `csvInspect` instead of `hasEmptyField` and `checkMissingFields` instead of `checkStructHasEmptyFields`

dweepgogia replied:
Done.

gallettilance reviewed (changes requested):
Great work! A couple comments mostly about names. Also please change the commit title to `[lib] - Handling mandatory and optional fields` or something similar and include a space between the title and the description.

dweepgogia commented:
Pushed the changes.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,12 +19,30 @@ func Verify(yamlFileName string) error {
 
-	fmt.Printf("%s is verified", yamlFileName)
+	// Flag for checking if there's any mandatory field missing.
+	checkError := false
```
Please rename to something more descriptive of the type. For example `isMandatoryFieldMissing` or `isErrorPresent`

dweepgogia replied:
Done.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
 
-	if err := yaml.Unmarshal(rawYAML, &csv); err != nil {
-		return nil, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	rawJson, err := yaml.YAMLToJSON(rawYAML)
```
why did you remove the old way of unmarshalling here?

dweepgogia replied:
Because we only have json tags in CSV.

gallettilance replied:
right. my bad!

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
+	default:
+		panic(fmt.Sprintf("%v kind is not supported.", v.Kind()))
+	}
 }
```
Nit: please add a new line at the end of the file

dweepgogia replied:
Done.

gallettilance replied:
did you remove the new line?

dweepgogia replied:
I still have it on my local. Btw what do we need it for?

gallettilance replied:
ok ok it's just hard to tell in this UI

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
```
This seems wrong. What if the value is actually 0?

dweepgogia replied:
Addressed this in a comment.

dweepgogia commented:
Pushed changes. Please review.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,12 +21,35 @@ func Verify(yamlFileName string) error {
+
+	if isMandatoryFieldMissing {
+		fmt.Printf("Populate all the mandatory fields missing from %s file.", yamlFileName)
+		os.Exit(1)
```
if you are going to exit the function without returning an `error`, might as well remove that from the function definition and have it not return anything.

dweepgogia replied:
Not making this change in the CLI in this PR.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,12 +21,35 @@ func Verify(yamlFileName string) error {
+		fmt.Printf("Populate all the mandatory fields missing from %s file.", yamlFileName)
+		os.Exit(1)
+	} else {
+		fmt.Printf("%s is verified.", yamlFileName)
```
then here you would just exit(0) I guess

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -31,14 +58,128 @@ func Verify(yamlFileName string) error {
+				continue
+			}
+			log = checkMissingFields(fieldValue, newParent, log)
+		default:
```
most of this code is redundant. Maybe you should do the empty check first and then do the switch statement on the kind to see if you have to recursively call the function

dweepgogia replied:
Done.

dweepgogia commented:
Pushed the changes. Please review.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
+	// Contains error logs for all missing optional and mandatory fields.
+	errorLog := csvInspect(csv)
+	checkManifestResult(errorLog.warnings)
+	checkManifestResult(errorLog.errors)
```
can we put this in the if statement below?

dweepgogia replied:
Do you want to check for `nil` before calling `checkManifestResult`?

gallettilance replied:
I just think since you are printing the errors only if it's not nil, and you are raising an error only if its not nil, might as well include all of that under the same IF no?

dweepgogia replied:
Done.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
+}
+
+// Iterates over the list of warnings and errors.
+func checkManifestResult(err []missingTypeError) {
```
not sure this is the right name for what this is doing

dweepgogia replied:
How about `getErrorsFromManifestResult`?

gallettilance replied:
yeah!

dweepgogia replied:
Done.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
+}
+
+// Asserts type to get the underlying field value.
+func assertTypeToGetValue(v interface{}) {
```
this name is strange too

dweepgogia replied:
Well, the only thing this function is doing is asserting type and getting the value.

gallettilance replied:
But the getting is more of a print no?

dweepgogia replied:
The way we describe the if statement in line `67` is it is type checking and getting the underlying struct value. See https://golangbot.com/interfaces-part-1/#typeassertion.

gallettilance replied:
ah ok gotcha thanks for clarifying!

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
+	if emptyVal && isOptionalField {
+		errString := fmt.Sprintf("Warning: Optional %s Missing.", typeName)
+		log.warnings = append(log.warnings, missingTypeError{errString, typeName, newParentStructName, false})
+		//log = append(log, fmt.Sprintf("WARNING: Optional %s missing (%s).", typeName, newParentStructName))
```
Please remove commented out code

dweepgogia replied:
Done.

gallettilance reviewed (commented):
Looks good just a few minor things

dweepgogia commented:
Pushed the changes. Please review.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -17,28 +33,181 @@ func Verify(yamlFileName string) error {
+// Iterates over the list of warnings and errors.
+func getErrorsFromManifestResult(err []missingTypeError) {
+	if err != nil {
+		for _, v := range err {
```
what happens if err is nil here?

dweepgogia replied:
No error/warning gets printed. Control goes back and you just see `csv.yaml file is verified.`

dweepgogia replied:
Do I need to explicitly `return` at the end if err is nil?

dweepgogia committed:
[doc] README update

- To add information about support for handling missing mandatory/optional fields.
